File: Python/schronisko.py
from dataclasses import dataclass
from datetime import datetime
from typing import Optional, List


@dataclass
class Animal:
    id: int
    name: str
    species: str
    age: int
    adopted: bool = False
    volunteer_id: Optional[int] = None

    def __str__(self) -> str:
        guardian = f"opiekun ID {self.volunteer_id}" if self.volunteer_id else "brak opiekuna"
        status = "zaadoptowane" if self.adopted else "dostępne"
        return f"[{self.id}] {self.name} ({self.species}, {self.age} lat) | {status} | {guardian}"


@dataclass
class Volunteer:
    id: int
    name: str
    phone: str
    assigned_animals: List[int]

    def __str__(self) -> str:
        animals = ", ".join(map(str, self.assigned_animals)) if self.assigned_animals else "brak"
        return f"[{self.id}] {self.name}, tel. {self.phone} | zwierzęta: {animals}"


@dataclass
class Adoption:
    id: int
    animal_id: int
    animal_name: str
    adopter_name: str
    contact: str
    processed_by: str
    date: str

    def __str__(self) -> str:
        return (
            f"[{self.id}] {self.date} | {self.animal_name} (ID {self.animal_id}) "
            f"| adoptujący: {self.adopter_name}, kontakt: {self.contact}, "
            f"obsłużył: {self.processed_by}"
        )


class Shelter:
    def __init__(self) -> None:
        self.animals: List[Animal] = []
        self.volunteers: List[Volunteer] = []
        self.adoptions: List[Adoption] = []
        self.next_animal_id = 1
        self.next_volunteer_id = 1
        self.next_adoption_id = 1

    def add_animal(self, name: str, species: str, age: int) -> None:
        self.animals.append(Animal(self.next_animal_id, name, species, age))
        self.next_animal_id += 1

    def add_volunteer(self, name: str, phone: str) -> None:
        self.volunteers.append(Volunteer(self.next_volunteer_id, name, phone, []))
        self.next_volunteer_id += 1

    def get_animal(self, animal_id: int) -> Optional[Animal]:
        return next((a for a in self.animals if a.id == animal_id), None)

    def get_volunteer(self, volunteer_id: int) -> Optional[Volunteer]:
        return next((v for v in self.volunteers if v.id == volunteer_id), None)

    def assign_volunteer(self, animal_id: int, volunteer_id: int) -> str:
        animal = self.get_animal(animal_id)
        volunteer = self.get_volunteer(volunteer_id)

        if animal is None:
            return "Nie znaleziono zwierzęcia."
        if volunteer is None:
            return "Nie znaleziono wolontariusza."
        if animal.adopted:
            return "Nie można przypisać opiekuna do zwierzęcia po adopcji."

        animal.volunteer_id = volunteer.id
        if animal.id not in volunteer.assigned_animals:
            volunteer.assigned_animals.append(animal.id)
        return "Opiekun został przypisany."

    def adopt_animal(self, animal_id: int, adopter_name: str, contact: str) -> str:
        animal = self.get_animal(animal_id)
        if animal is None:
            return "Nie znaleziono zwierzęcia."
        if animal.adopted:
            return "To zwierzę jest już adoptowane."

        processed_by = "system"
        if animal.volunteer_id is not None:
            volunteer = self.get_volunteer(animal.volunteer_id)
            if volunteer:
                processed_by = volunteer.name

        animal.adopted = True
        adoption = Adoption(
            self.next_adoption_id,
            animal.id,
            animal.name,
            adopter_name,
            contact,
            processed_by,
            datetime.now().strftime("%Y-%m-%d %H:%M")
        )
        self.adoptions.append(adoption)
        self.next_adoption_id += 1
        return f"Zwierzę '{animal.name}' zostało zaadoptowane."

    def show_animals(self) -> None:
        if not self.animals:
            print("Brak zwierząt.")
            return
        for animal in self.animals:
            print(animal)

    def show_volunteers(self) -> None:
        if not self.volunteers:
            print("Brak wolontariuszy.")
            return
        for volunteer in self.volunteers:
            print(volunteer)

    def show_adoptions(self) -> None:
        if not self.adoptions:
            print("Brak adopcji.")
            return
        for adoption in self.adoptions:
            print(adoption)


def read_int(prompt: str) -> int:
    while True:
        try:
            return int(input(prompt))
        except ValueError:
            print("Podaj poprawną liczbę całkowitą.")


def main() -> None:
    shelter = Shelter()

    while True:
        print("\n=== SCHRONISKO DLA ZWIERZĄT ===")
        print("1. Dodaj zwierzę")
        print("2. Dodaj wolontariusza")
        print("3. Pokaż zwierzęta")
        print("4. Pokaż wolontariuszy")
        print("5. Przydziel opiekuna do zwierzęcia")
        print("6. Przeprowadź adopcję")
        print("7. Pokaż adopcje")
        print("0. Wyjście")

        choice = input("Wybierz opcję: ").strip()

        if choice == "1":
            name = input("Imię zwierzęcia: ")
            species = input("Gatunek: ")
            age = read_int("Wiek: ")
            shelter.add_animal(name, species, age)
            print("Dodano zwierzę.")

        elif choice == "2":
            name = input("Imię i nazwisko wolontariusza: ")
            phone = input("Telefon: ")
            shelter.add_volunteer(name, phone)
            print("Dodano wolontariusza.")

        elif choice == "3":
            shelter.show_animals()

        elif choice == "4":
            shelter.show_volunteers()

        elif choice == "5":
            animal_id = read_int("ID zwierzęcia: ")
            volunteer_id = read_int("ID wolontariusza: ")
            print(shelter.assign_volunteer(animal_id, volunteer_id))

        elif choice == "6":
            animal_id = read_int("ID zwierzęcia: ")
            adopter_name = input("Imię i nazwisko adoptującego: ")
            contact = input("Kontakt: ")
            print(shelter.adopt_animal(animal_id, adopter_name, contact))

        elif choice == "7":
            shelter.show_adoptions()

        elif choice == "0":
            print("Koniec programu.")
            break

        else:
            print("Niepoprawna opcja.")


if __name__ == "__main__":
    main()

File: Csharp/schronisko.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class Animal
{
    public int Id { get; set; }
    public string Name { get; set; } = "";
    public string Species { get; set; } = "";
    public int Age { get; set; }
    public bool Adopted { get; set; }
    public int? VolunteerId { get; set; }

    public override string ToString()
    {
        string guardian = VolunteerId.HasValue ? $"opiekun ID {VolunteerId.Value}" : "brak opiekuna";
        string status = Adopted ? "zaadoptowane" : "dostępne";
        return $"[{Id}] {Name} ({Species}, {Age} lat) | {status} | {guardian}";
    }
}

public class Volunteer
{
    public int Id { get; set; }
    public string Name { get; set; } = "";
    public string Phone { get; set; } = "";
    public List<int> AssignedAnimals { get; set; } = new List<int>();

    public override string ToString()
    {
        string animals = AssignedAnimals.Count > 0 ? string.Join(", ", AssignedAnimals) : "brak";
        return $"[{Id}] {Name}, tel. {Phone} | zwierzęta: {animals}";
    }
}

public class Adoption
{
    public int Id { get; set; }
    public int AnimalId { get; set; }
    public string AnimalName { get; set; } = "";
    public string AdopterName { get; set; } = "";
    public string Contact { get; set; } = "";
    public string ProcessedBy { get; set; } = "";
    public DateTime Date { get; set; }

    public override string ToString()
    {
        return $"[{Id}] {Date:yyyy-MM-dd HH:mm} | {AnimalName} (ID {AnimalId}) | adoptujący: {AdopterName}, kontakt: {Contact}, obsłużył: {ProcessedBy}";
    }
}

public class Shelter
{
    private readonly List<Animal> animals = new List<Animal>();
    private readonly List<Volunteer> volunteers = new List<Volunteer>();
    private readonly List<Adoption> adoptions = new List<Adoption>();

    private int nextAnimalId = 1;
    private int nextVolunteerId = 1;
    private int nextAdoptionId = 1;

    public void AddAnimal(string name, string species, int age)
    {
        animals.Add(new Animal
        {
            Id = nextAnimalId++,
            Name = name,
            Species = species,
            Age = age,
            Adopted = false,
            VolunteerId = null
        });
    }

    public void AddVolunteer(string name, string phone)
    {
        volunteers.Add(new Volunteer
        {
            Id = nextVolunteerId++,
            Name = name,
            Phone = phone
        });
    }

    private Animal? GetAnimal(int id) => animals.FirstOrDefault(a => a.Id == id);
    private Volunteer? GetVolunteer(int id) => volunteers.FirstOrDefault(v => v.Id == id);

    public string AssignVolunteer(int animalId, int volunteerId)
    {
        Animal? animal = GetAnimal(animalId);
        Volunteer? volunteer = GetVolunteer(volunteerId);

        if (animal == null) return "Nie znaleziono zwierzęcia.";
        if (volunteer == null) return "Nie znaleziono wolontariusza.";
        if (animal.Adopted) return "Nie można przypisać opiekuna do zwierzęcia po adopcji.";

        animal.VolunteerId = volunteer.Id;
        if (!volunteer.AssignedAnimals.Contains(animal.Id))
        {
            volunteer.AssignedAnimals.Add(animal.Id);
        }

        return "Opiekun został przypisany.";
    }

    public string AdoptAnimal(int animalId, string adopterName, string contact)
    {
        Animal? animal = GetAnimal(animalId);

        if (animal == null) return "Nie znaleziono zwierzęcia.";
        if (animal.Adopted) return "To zwierzę jest już adoptowane.";

        string processedBy = "system";
        if (animal.VolunteerId.HasValue)
        {
            Volunteer? volunteer = GetVolunteer(animal.VolunteerId.Value);
            if (volunteer != null)
            {
                processedBy = volunteer.Name;
            }
        }

        animal.Adopted = true;

        adoptions.Add(new Adoption
        {
            Id = nextAdoptionId++,
            AnimalId = animal.Id,
            AnimalName = animal.Name,
            AdopterName = adopterName,
            Contact = contact,
            ProcessedBy = processedBy,
            Date = DateTime.Now
        });

        return $"Zwierzę '{animal.Name}' zostało zaadoptowane.";
    }

    public void ShowAnimals()
    {
        if (animals.Count == 0)
        {
            Console.WriteLine("Brak zwierząt.");
            return;
        }

        foreach (Animal animal in animals)
        {
            Console.WriteLine(animal);
        }
    }

    public void ShowVolunteers()
    {
        if (volunteers.Count == 0)
        {
            Console.WriteLine("Brak wolontariuszy.");
            return;
        }

        foreach (Volunteer volunteer in volunteers)
        {
            Console.WriteLine(volunteer);
        }
    }

    public void ShowAdoptions()
    {
        if (adoptions.Count == 0)
        {
            Console.WriteLine("Brak adopcji.");
            return;
        }

        foreach (Adoption adoption in adoptions)
        {
            Console.WriteLine(adoption);
        }
    }
}

public class Program
{
    private static int ReadInt(string prompt)
    {
        while (true)
        {
            Console.Write(prompt);
            string? input = Console.ReadLine();

            if (int.TryParse(input, out int value))
            {
                return value;
            }

            Console.WriteLine("Podaj poprawną liczbę całkowitą.");
        }
    }

    public static void Main()
    {
        Shelter shelter = new Shelter();

        while (true)
        {
            Console.WriteLine();
            Console.WriteLine("=== SCHRONISKO DLA ZWIERZĄT ===");
            Console.WriteLine("1. Dodaj zwierzę");
            Console.WriteLine("2. Dodaj wolontariusza");
            Console.WriteLine("3. Pokaż zwierzęta");
            Console.WriteLine("4. Pokaż wolontariuszy");
            Console.WriteLine("5. Przydziel opiekuna do zwierzęcia");
            Console.WriteLine("6. Przeprowadź adopcję");
            Console.WriteLine("7. Pokaż adopcje");
            Console.WriteLine("0. Wyjście");

            Console.Write("Wybierz opcję: ");
            string? choice = Console.ReadLine();

            switch (choice)
            {
                case "1":
                    Console.Write("Imię zwierzęcia: ");
                    string animalName = Console.ReadLine() ?? "";
                    Console.Write("Gatunek: ");
                    string species = Console.ReadLine() ?? "";
                    int age = ReadInt("Wiek: ");
                    shelter.AddAnimal(animalName, species, age);
                    Console.WriteLine("Dodano zwierzę.");
                    break;

                case "2":
                    Console.Write("Imię i nazwisko wolontariusza: ");
                    string volunteerName = Console.ReadLine() ?? "";
                    Console.Write("Telefon: ");
                    string phone = Console.ReadLine() ?? "";
                    shelter.AddVolunteer(volunteerName, phone);
                    Console.WriteLine("Dodano wolontariusza.");
                    break;

                case "3":
                    shelter.ShowAnimals();
                    break;

                case "4":
                    shelter.ShowVolunteers();
                    break;

                case "5":
                    int animalIdForVolunteer = ReadInt("ID zwierzęcia: ");
                    int volunteerId = ReadInt("ID wolontariusza: ");
                    Console.WriteLine(shelter.AssignVolunteer(animalIdForVolunteer, volunteerId));
                    break;

                case "6":
                    int animalIdForAdoption = ReadInt("ID zwierzęcia: ");
                    Console.Write("Imię i nazwisko adoptującego: ");
                    string adopterName = Console.ReadLine() ?? "";
                    Console.Write("Kontakt: ");
                    string contact = Console.ReadLine() ?? "";
                    Console.WriteLine(shelter.AdoptAnimal(animalIdForAdoption, adopterName, contact));
                    break;

                case "7":
                    shelter.ShowAdoptions();
                    break;

                case "0":
                    Console.WriteLine("Koniec programu.");
                    return;

                default:
                    Console.WriteLine("Niepoprawna opcja.");
                    break;
            }
        }
    }
}